### PR TITLE
Fix guided decoding crashes

### PR DESCRIPTION
**Pull request by kzawora-intel**
This PR mostly ports https://github.com/vllm-project/vllm/pull/11389 to the design introduced by https://github.com/HabanaAI/vllm-fork/pull/358 and makes the custom caching code a little bit more robust.
Currently there are two problems with guided decode:
- `mask[list(allowed_tokens)] = 0` is causing crashes due to allowed_tokens containing tensors. Pretty easy fix.
- The value type of `self._fsm_state` was changed from `int` to union of `int` and `outlines.state.CFGState`, which may cause `self._cached_get_mask_tensor(state_id, scores.size(-1), scores.device)` to crash, as `outlines.state.CFGState` is not hashable. This PR changes the caching mechanism so that if function arguments are not hashable, their id is taken as key. This might cause some cache misses, but that's better than crashing, as it does right now.
None of the above is problem on upstream, as this stems from code introduced in https://github.com/HabanaAI/vllm-fork/pull/358.
I've also added guided decode tests to CI suite.

### CODE REVIEW

File: vllm/model_executor/guided_decoding/outlines_logits_processors.py [hunk 1]
```diff
@@ -60,7 +90,15 @@ def __init__(self, guide: Guide):
     @lru_cache(maxsize=128)
     def _create_mask_tensor(allowed_tokens, vocab_size, device):
         mask = torch.full((vocab_size, ), -math.inf, device=device)
-        mask[list(allowed_tokens)] = 0
+        # The tokenizer may support more token ids than the model can generate,
+        # eg. Llama 3.2 Vision models have an `<|image|>` token with id 128256
+        # but scores.shape == torch.Size([128256])
+        # Using NumPy is faster for filtering token ids
+        allowed_tokens = np.array(allowed_tokens, dtype=np.int64)
+        allowed_tokens = torch.tensor(allowed_tokens, device=device)
```
Comment (madamczykhabana) [inline vllm/model_executor/guided_decoding/outlines_logits_processors.py:98]: Why we're first converting to np.array and then to torch.tensor? Can't we create the tensor directly?

File: vllm/model_executor/guided_decoding/outlines_logits_processors.py [hunk 2]
```diff
@@ -60,7 +90,15 @@ def __init__(self, guide: Guide):
     @lru_cache(maxsize=128)
     def _create_mask_tensor(allowed_tokens, vocab_size, device):
         mask = torch.full((vocab_size, ), -math.inf, device=device)
-        mask[list(allowed_tokens)] = 0
+        # The tokenizer may support more token ids than the model can generate,
+        # eg. Llama 3.2 Vision models have an `<|image|>` token with id 128256
+        # but scores.shape == torch.Size([128256])
+        # Using NumPy is faster for filtering token ids
```
Comment (madamczykhabana) [inline vllm/model_executor/guided_decoding/outlines_logits_processors.py:96]: You say that numpy is faster, but you're operating on tensors (masked_select + index_fill)

File: vllm/model_executor/guided_decoding/outlines_logits_processors.py [hunk 3]
```diff
@@ -60,7 +90,15 @@ def __init__(self, guide: Guide):
...
+        allowed_tokens = np.array(allowed_tokens, dtype=np.int64)
+        allowed_tokens = torch.tensor(allowed_tokens, device=device)
+        allowed_tokens = allowed_tokens.masked_select(
```
Comment (madamczykhabana) [inline vllm/model_executor/guided_decoding/outlines_logits_processors.py:99]: If I'm not mistaken masked_select might introduce dynamicity on HPU and cause recompilations. What's the device here? CPU or HPU?

File: vllm/model_executor/guided_decoding/outlines_logits_processors.py [hunk 4]
```diff
@@ -36,12 +38,40 @@
...
+                  - Returns True if the object is an instance of Hashable.
+                  - If the object is an Iterable, returns True only if all items
+                  in the iterable are hashable.
+                  - Returns False for all other cases.
+        """
+        match obj:
+            case CFGState():
+                return False
+            case Iterable():
+                is_all_hashable = True
```
Comment (madamczykhabana) [inline vllm/model_executor/guided_decoding/outlines_logits_processors.py:59]: `all(is_hashable(sub_item) for sub_item in item)`

File: vllm/model_executor/guided_decoding/outlines_logits_processors.py [hunk 5]
```diff
@@ -36,12 +38,40 @@
 def _cached(fn):
     cache: Dict[Any, Any] = {}
 
+    def is_hashable(obj):
```
Comment (madamczykhabana) [inline vllm/model_executor/guided_decoding/outlines_logits_processors.py:41]: Hmm, I'm thinking whether we should take a slightly different approach. Let me give you an example:
foo = [1, CFGState]
is_hashable(foo) = false => key(foo) = id(foo)
but if we change the logic so that instead of checking if everything is hashable we hash everything we can and use id where it's not possible? i.e.
semi_hash(obj) =
  case Iterable -> hash(semi_hash(sub) for sub in obj)
  case CFGState -> id(obj)
  case Hashable -> hash(obj)
This way we could hash foo like this:
semi_hash(foo) = hash(hash(int), id(CFGState))

File: vllm/model_executor/guided_decoding/outlines_logits_processors.py (hunk 1 above)
Comment (kzawora-intel) [reply vllm/model_executor/guided_decoding/outlines_logits_processors.py:98]: this is actually taken straight from https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/guided_decoding/outlines_logits_processors.py#L85-L94 and I agree it's a bit silly, I'll remove that

File: vllm/model_executor/guided_decoding/outlines_logits_processors.py (hunk 3 above)
Comment (kzawora-intel) [reply vllm/model_executor/guided_decoding/outlines_logits_processors.py:99]: this is going to use whatever device we used in previous design - by default on hpu we'll do that on CPU

File: vllm/model_executor/guided_decoding/outlines_logits_processors.py (hunk 2 above)
Comment (kzawora-intel) [reply vllm/model_executor/guided_decoding/outlines_logits_processors.py:96]: same as the np.array conversion stuff - this is straight out of upstream, and it does seem kinda silly

File: vllm/model_executor/guided_decoding/outlines_logits_processors.py (hunk 5 above)
Comment (kzawora-intel) [reply vllm/model_executor/guided_decoding/outlines_logits_processors.py:41]: I adopted this approach in the new `hash_args` function, but I decided not to make a special case for `CFGState`, as it's an iterable (namedtuple) containing some non-hashable stuff - so now it's something like
```
hash_args(obj) =
  case Iterable -> hash(tuple(hash_args(sub) for sub in obj))
  case Hashable -> hash(obj)
  case _ -> hash(id(obj))
```
so with `foo = [1, CFGState]`
the hash would look like this:
`semi_hash(foo) = hash(hash(int), hash(tuple(id(PartialParserState), hash(tuple(int, int, ...)))`

the hasher now has one drawback that is currently a non-issue - namely if `CFGState` object share the same `PartialParserState` and the state is modified with each call, the id (and thus the hash) will stay the same, but the contents aren't - fortunately for us, it seems like the second field of CFGState is going to give that away; if we encounter bugs here, it probably would be a better idea to use hashes of some random uuids rather than unhashable object id

File: vllm/model_executor/guided_decoding/outlines_logits_processors.py (hunk 4 above)
Comment (kzawora-intel) [reply vllm/model_executor/guided_decoding/outlines_logits_processors.py]: the function got removed

**Review by madamczykhabana — APPROVED** [review]
LGTM
